### PR TITLE
Added an optional dtype argument to BatchNorm.

### PR DESCRIPTION
**Pull request by Benjamin-Walker**
Was encountering the error "Old and new values have different structures." when using BatchNorm with complex valued inputs. Added the ability to specify the dtype of the buffers passed to StateIndex. 

I am slightly unsure on the typing to use for dtype (as using jnp.dtype raises a style error during pyright's pre-commit hook). I have currently opted for Num from jaxtyping.

### CODE REVIEW

File: equinox/nn/_batch_norm.py [hunk 1]
```diff
@@ -60,6 +60,7 @@ def __init__(
         channelwise_affine: bool = True,
         momentum: float = 0.99,
         inference: bool = False,
+        dtype: Num = jnp.float32,
```
Comment (patrick-kidger) [inline equinox/nn/_batch_norm.py:63]: `Num` is only allowed to be used in conjunction with an array and shape, e.g. `Num[Array, "foo"]`.

Surprisingly, there actually isn't a valid dtype annotation that I know of. For now I'd just leave this unannotated.

File: equinox/nn/_batch_norm.py [hunk 2]
```diff
@@ -139,7 +144,7 @@ def __call__(
             def _stats(y):
                 mean = jnp.mean(y)
                 mean = lax.pmean(mean, self.axis_name)
-                var = jnp.mean((y - mean) ** 2)
+                var = jnp.mean(abs(y - mean) ** 2)
```
Comment (patrick-kidger) [inline equinox/nn/_batch_norm.py:147]: Slightly better would be `(y - mean) * jnp.conj(y - mean)`.

**Comment by Benjamin-Walker** [comment]
Hey! No worries, made the changes and rebased into a single commit locally, but still not quite got the hang of doing so such that it updates the pull request properly! Played about a bit too much and going to need to open a new pull request now, apologies!

**Comment by patrick-kidger** [comment]
Haha sure. FWIW the commands for this are something like `git reset --soft HEAD~5; git commit; git push -f`

**Comment by Benjamin-Walker** [comment]
Thank you, will keep a note of that somewhere. Was trying to use `git rebase -i HEAD~5`, squashing all my commits into one, and then `git push -f`. Updated my forked branch but didn't transfer across to this pull request.
